File: 2024/day23/perl/solution.pl
#!/usr/bin/env perl
use strict;
use warnings;
use List::Util qw(max);

sub parse_input {
    my ($filename) = @_;
    my %graph;

    open my $fh, '<', $filename or die "Cannot open $filename: $!";
    while (my $line = <$fh>) {
        chomp $line;
        my ($a, $b) = split /-/, $line;
        $graph{$a}{$b} = 1;
        $graph{$b}{$a} = 1;
    }
    close $fh;

    return \%graph;
}

sub find_triangles {
    my ($graph) = @_;
    my %triangles;

    for my $a (keys %$graph) {
        for my $b (keys %{$graph->{$a}}) {
            next unless $a lt $b;  # Only process each edge once

            # Find common neighbors (intersection of neighbors of a and b)
            for my $c (keys %{$graph->{$a}}) {
                if (exists $graph->{$b}{$c}) {
                    # Create sorted triangle key
                    my @tri = sort ($a, $b, $c);
                    my $key = join(',', @tri);
                    $triangles{$key} = \@tri;
                }
            }
        }
    }

    return \%triangles;
}

sub part1 {
    my ($graph) = @_;
    my $triangles = find_triangles($graph);
    my $count = 0;

    for my $key (keys %$triangles) {
        my $tri = $triangles->{$key};
        if (grep { /^t/ } @$tri) {
            $count++;
        }
    }

    return $count;
}

sub set_intersection {
    my ($set1, $set2) = @_;
    my %result;
    for my $key (keys %$set1) {
        $result{$key} = 1 if exists $set2->{$key};
    }
    return \%result;
}

sub set_union {
    my ($set1, $set2) = @_;
    my %result = %$set1;
    for my $key (keys %$set2) {
        $result{$key} = 1;
    }
    return \%result;
}

sub set_difference {
    my ($set1, $set2) = @_;
    my %result;
    for my $key (keys %$set1) {
        $result{$key} = 1 unless exists $set2->{$key};
    }
    return \%result;
}

sub bron_kerbosch {
    my ($graph, $r, $p, $x, $cliques) = @_;

    # If both P and X are empty, R is a maximal clique
    if (!%$p && !%$x) {
        push @$cliques, {%$r};
        return;
    }

    # Choose pivot from P ∪ X to minimize branching
    my $p_union_x = set_union($p, $x);
    my $pivot = '';
    my $max_neighbors = -1;

    for my $v (keys %$p_union_x) {
        my $neighbors_in_p = set_intersection($graph->{$v}, $p);
        my $count = scalar keys %$neighbors_in_p;
        if ($count > $max_neighbors) {
            $max_neighbors = $count;
            $pivot = $v;
        }
    }

    # For each vertex in P \ N(pivot)
    my $pivot_neighbors = exists $graph->{$pivot} ? $graph->{$pivot} : {};
    my $candidates = set_difference($p, $pivot_neighbors);

    for my $v (keys %$candidates) {
        my $new_r = set_union($r, {$v => 1});
        my $new_p = set_intersection($p, $graph->{$v});
        my $new_x = set_intersection($x, $graph->{$v});

        bron_kerbosch($graph, $new_r, $new_p, $new_x, $cliques);

        delete $p->{$v};
        $x->{$v} = 1;
    }
}

sub part2 {
    my ($graph) = @_;
    my @cliques;

    # Initialize P with all nodes
    my %all_nodes = map { $_ => 1 } keys %$graph;

    bron_kerbosch($graph, {}, \%all_nodes, {}, \@cliques);

    # Find the largest clique
    my $largest = $cliques[0];
    my $max_size = scalar keys %$largest;

    for my $clique (@cliques) {
        my $size = scalar keys %$clique;
        if ($size > $max_size) {
            $max_size = $size;
            $largest = $clique;
        }
    }

    # Return sorted, comma-joined password
    my @sorted = sort keys %$largest;
    return join(',', @sorted);
}

sub main {
    my $graph = parse_input('../input.txt');

    print 'Part 1: ', part1($graph), "\n";
    print 'Part 2: ', part2($graph), "\n";
}

main();

File: 2024/day23/php/solution.php
#!/usr/bin/env php
<?php

function parseInput($filename) {
    $graph = [];
    $lines = file($filename, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);

    foreach ($lines as $line) {
        list($a, $b) = explode('-', trim($line));

        if (!isset($graph[$a])) {
            $graph[$a] = [];
        }
        if (!isset($graph[$b])) {
            $graph[$b] = [];
        }

        $graph[$a][$b] = true;
        $graph[$b][$a] = true;
    }

    return $graph;
}

function findTriangles($graph) {
    $triangles = [];

    foreach ($graph as $a => $neighborsA) {
        foreach ($neighborsA as $b => $_) {
            if ($a < $b) {
                // Find common neighbors
                foreach ($neighborsA as $c => $_) {
                    if (isset($graph[$b][$c]) && $a < $c && $b < $c) {
                        // Create sorted triangle key
                        $tri = [$a, $b, $c];
                        sort($tri);
                        $key = implode(',', $tri);
                        $triangles[$key] = $tri;
                    }
                }
            }
        }
    }

    return array_values($triangles);
}

function part1($graph) {
    $triangles = findTriangles($graph);
    $count = 0;

    foreach ($triangles as $tri) {
        foreach ($tri as $node) {
            if ($node[0] === 't') {
                $count++;
                break;
            }
        }
    }

    return $count;
}

function bronKerbosch($graph, $r, $p, $x, &$cliques) {
    if (empty($p) && empty($x)) {
        $cliques[] = $r;
        return;
    }

    // Find pivot with maximum neighbors in p
    $union = array_merge($p, $x);
    $maxNeighbors = -1;
    $pivot = null;

    foreach ($union as $v) {
        $neighbors = isset($graph[$v]) ? array_intersect_key($graph[$v], $p) : [];
        $count = count($neighbors);
        if ($count > $maxNeighbors) {
            $maxNeighbors = $count;
            $pivot = $v;
        }
    }

    // Get pivot neighbors in p
    $pivotNeighbors = isset($graph[$pivot]) ? array_intersect_key($graph[$pivot], $p) : [];
    $candidates = array_diff_key($p, $pivotNeighbors);

    foreach ($candidates as $v => $_) {
        $neighbors = isset($graph[$v]) ? $graph[$v] : [];

        $newR = $r;
        $newR[$v] = true;

        $newP = array_intersect_key($p, $neighbors);
        $newX = array_intersect_key($x, $neighbors);

        bronKerbosch($graph, $newR, $newP, $newX, $cliques);

        unset($p[$v]);
        $x[$v] = true;
    }
}

function part2($graph) {
    $cliques = [];
    $allNodes = [];

    foreach ($graph as $node => $_) {
        $allNodes[$node] = true;
    }

    bronKerbosch($graph, [], $allNodes, [], $cliques);

    // Find the largest clique
    $largest = [];
    foreach ($cliques as $clique) {
        if (count($clique) > count($largest)) {
            $largest = $clique;
        }
    }

    // Return sorted, comma-joined password
    $nodes = array_keys($largest);
    sort($nodes);
    return implode(',', $nodes);
}

function main() {
    $graph = parseInput('../input.txt');

    echo 'Part 1: ' . part1($graph) . "\n";
    echo 'Part 2: ' . part2($graph) . "\n";
}

main();
